File: src/main/kotlin/magicbook/gtlitecore/common/item/behavior/BerryCropSeedBehavior.java
package magicbook.gtlitecore.common.item.behavior;

import gregtech.api.items.metaitem.MetaItem;
import magicbook.gtlitecore.common.block.blocks.GTLiteBerryBushVariantBlock;
import net.minecraft.client.resources.I18n;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemStack;
import net.minecraft.util.ActionResult;
import net.minecraft.util.EnumActionResult;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.EnumHand;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.text.TextComponentTranslation;
import net.minecraft.world.World;

import java.util.List;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicReference;

public class BerryCropSeedBehavior extends CropSeedBehavior
{

    public BerryCropSeedBehavior(GTLiteBerryBushVariantBlock berryBush, ItemStack seed, ItemStack crop)
    {
        super(berryBush, seed, crop);
    }

    public BerryCropSeedBehavior(GTLiteBerryBushVariantBlock berryBush, MetaItem<?>.MetaValueItem seed, MetaItem<?>.MetaValueItem crop)
    {
        super(berryBush, seed, crop);
    }

    @Override
    public ActionResult<ItemStack> onItemUse(EntityPlayer player, World worldIn, BlockPos pos,
                                             EnumHand hand, EnumFacing facing,
                                             float hitX, float hitY, float hitZ)
    {
        if (!isBlocked(worldIn, pos, player))
            return super.onItemUse(player, worldIn, pos, hand, facing, hitX, hitY, hitZ);
        return new ActionResult<>(EnumActionResult.FAIL, player.getHeldItem(hand));
    }

    @Override
    public void addInformation(ItemStack stack, List<String> lines)
    {
        super.addInformation(stack, lines);
        lines.add(I18n.format("gtlitecore.tooltip.crop.berry_bush.1"));
        lines.add(I18n.format("gtlitecore.tooltip.crop.berry_bush.2"));
        lines.add(I18n.format("gtlitecore.tooltip.crop.berry_bush.3"));
        lines.add(I18n.format("gtlitecore.tooltip.crop.berry_bush.4"));
    }

    private boolean isBlocked(World world, BlockPos pos, EntityPlayer player)
    {
        AtomicInteger areAnyBlocked = new AtomicInteger(0);
        AtomicReference<BlockPos> blockedCrop = new AtomicReference<>();
        BlockPos.getAllInBox(pos.up().east().north(), pos.up().west().south()).forEach((crop) -> {
            if (crop.equals(pos.up()) || world.getBlockState(crop).getBlock() instanceof GTLiteBerryBushVariantBlock)
            {
                AtomicBoolean isBlocked = new AtomicBoolean(true);
                BlockPos.getAllInBox(crop.east().north(), crop.west().south()).forEach((blockpos) -> {
                    if (!blockpos.equals(pos.up()) && world.getBlockState(blockpos).getBlock().isAir(world.getBlockState(blockpos), world, blockpos))
                        isBlocked.set(false);
                });
                if (isBlocked.get())
                {
                    blockedCrop.set(crop);
                    areAnyBlocked.set(areAnyBlocked.get() + 1);
                }
            }
        });
        if (world.isRemote && areAnyBlocked.get() > 0)
        {
            String posString = "(" + blockedCrop.get().getX() + ", " + blockedCrop.get().getY() + ", " + blockedCrop.get().getZ() + ")";
            player.sendMessage(new TextComponentTranslation("gtlitecore.message.berry_bush_blocked", posString, areAnyBlocked.get()));
        }
        return areAnyBlocked.get() > 0;
    }
}

File: src/main/kotlin/gregtechlite/gtlitecore/common/item/behavior/BerryCropSeedBehavior.kt
package gregtechlite.gtlitecore.common.item.behavior

import gregtech.api.items.metaitem.MetaItem
import gregtechlite.gtlitecore.api.extension.stack
import gregtechlite.gtlitecore.common.block.GTLiteBerryBushBlock
import net.minecraft.client.resources.I18n
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.ItemStack
import net.minecraft.util.ActionResult
import net.minecraft.util.EnumActionResult
import net.minecraft.util.EnumFacing
import net.minecraft.util.EnumHand
import net.minecraft.util.math.BlockPos
import net.minecraft.util.text.TextComponentTranslation
import net.minecraft.world.World

class BerryCropSeedBehavior(berryBush: GTLiteBerryBushBlock, seed: ItemStack, crop: ItemStack) : CropSeedBehavior(berryBush, seed, crop)
{

    constructor(berryBush: GTLiteBerryBushBlock, seed: MetaItem<*>.MetaValueItem, crop: MetaItem<*>.MetaValueItem)
            : this(berryBush, seed.stack(), crop.stack())

    override fun onItemUse(player: EntityPlayer, worldIn: World, pos: BlockPos, hand: EnumHand, facing: EnumFacing?,
                           hitX: Float, hitY: Float, hitZ: Float): ActionResult<ItemStack?>?
    {
        if (!isBlocked(worldIn, pos, player))
        {
            return super.onItemUse(player, worldIn, pos, hand, facing, hitX, hitY, hitZ)
        }
        return ActionResult(EnumActionResult.FAIL, player.getHeldItem(hand))
    }

    override fun addInformation(stack: ItemStack, lines: MutableList<String>)
    {
        val cropType = "gtlitecore.tooltip.crop.type"
        val cropPrefix = "${cropType}.berry"
        lines.add(I18n.format(cropType) + I18n.format(cropPrefix))
    }

    private fun isBlocked(world: World, pos: BlockPos, player: EntityPlayer): Boolean
    {
        var blockedCount = 0
        var blockedCrop: BlockPos? = null

        //  |
        // [ ]
        val up = pos.up()

        //        [ ]
        //         |
        // [ ] -- [ ]
        //  |
        // [ ]
        val start = up.east().north()

        // [ ] -- [ ]
        //  |      |
        // [ ]    [ ]
        val end = up.west().south()

        for (crop in BlockPos.getAllInBox(start, end))
        {
            if (crop == up || world.getBlockState(crop).block is GTLiteBerryBushBlock)
            {
                var isBlocked = true
                val innerStart = crop.east().north()
                val innerEnd = crop.west().south()

                for (blockPos in BlockPos.getAllInBox(innerStart, innerEnd))
                {
                    // If the block state at pos is air, then marked it is not be blocked.
                    if (blockPos != up && world.getBlockState(blockPos)
                        .block.isAir(world.getBlockState(blockPos), world, blockPos))
                    {
                        isBlocked = false
                        break
                    }
                }
                // If the block state at pos is blocked, then update count.
                if (isBlocked)
                {
                    blockedCrop = crop
                    blockedCount++
                }
            }
        }

        // If the crop at pos is blocked with several blocks, then sending a message to player.
        if (world.isRemote && blockedCount > 0)
        {
            val cropPos = blockedCrop ?: up
            val cropPosInfo = "(${cropPos.x}, ${cropPos.y}, ${cropPos.z})"
            player.sendMessage(TextComponentTranslation(
                "gtlitecore.message.berry_bush_blocked", cropPosInfo, blockedCount))
        }

        return blockedCount > 0
    }

    override fun getDescription(): Array<String>
    {
        return arrayOf("gtlitecore.jei.berry_crop_info.1", "gtlitecore.jei.berry_crop_info.2",
                       "gtlitecore.jei.berry_crop_info.3", "gtlitecore.jei.berry_crop_info.4")
    }

}
